test/metabase/test_runner/assert_exprs/approximately_equal.clj

```diff
@@ -0,0 +1,270 @@
+    => true
+
+    (is (=? #approx [1.5 0.1]
+            1.6))
+    =>
+    expected: #approx [1.5 0.1]
+
+      actual: 1.6
+        diff: - (not (approx= 1.5 1.6 #_epsilon 0.1))
+              + nil"
+  (:require
+   [clojure.algo.generic.math-functions :as algo.generic.math]
+   [clojure.pprint :as pprint]
+   [methodical.core :as methodical]
+   [schema.core :as s]))
+
+(set! *warn-on-reflection* true)
+
+(methodical/defmulti ^:dynamic =?-diff
+  "Multimethod to use to diff two things with `=?`. Despite not having earmuffs, this is dynamic so it can be rebound at
+  runtime."
+  {:arglists '([expected actual])}
+  (fn [expected actual]
+    [(type expected) (type actual)]))
+
+(defn- add-primary-methods
+  "Add primary methods in map `m` of dispatch value -> method fn to [[*impl*]]. Return a new multifn with those methods
+  added."
+  [m]
+  (reduce
+   (fn [multifn [dispatch-value f]]
+     (methodical/add-primary-method multifn dispatch-value f))
+   =?-diff
+   m))
+
+(def ^:dynamic *debug*
+  "Whether to enable Methodical method tracing for debug purposes."
+  false)
+
+(defn =?-diff*
+  "Are `expected` and `actual` 'approximately' equal to one another?"
+  ([expected actual]
+   (=?-diff* =?-diff expected actual))
+
+  ([diff-fn expected actual]
+   (let [diff-fn (if (map? diff-fn)
+                   (add-primary-methods diff-fn)
+                   diff-fn)]
+     (binding [=?-diff diff-fn]
+       (if *debug*
+         (methodical/trace diff-fn expected actual)
+         (diff-fn expected actual))))))
+
+;;;; Default method impls
+
+(methodical/defmethod =?-diff :default
+  [expected actual]
+  (when-not (= expected actual)
+    (list 'not= expected actual)))
+
+(methodical/defmethod =?-diff [Class Object]
+  [expected-class actual]
+  (when-not (instance? expected-class actual)
+    (list 'not (list 'instance? expected-class actual))))
+
+(methodical/defmethod =?-diff [java.util.regex.Pattern String]
+  [expected-regex s]
+  (when-not (re-find expected-regex s)
+    (list 'not (list 're-find expected-regex s))))
+
+;;; two regexes should be treated as equal if they're the same pattern.
+(methodical/defmethod =?-diff [java.util.regex.Pattern java.util.regex.Pattern]
+  [expected actual]
+  (when-not (= (str expected) (str actual))
+    (list 'not= (list 'str expected) (list 'str actual))))
+
+(methodical/defmethod =?-diff [clojure.lang.AFunction Object]
+  [pred actual]
+  (when-not (pred actual)
+    (list 'not (list pred actual))))
+
+(methodical/defmethod =?-diff [clojure.lang.Sequential clojure.lang.Sequential]
+  [expected-seq actual-seq]
+  (loop [acc                        []
+         [expected & more-expected] expected-seq
+         [actual & more-actual]     actual-seq]
+    (let [diff (=?-diff expected actual)
+          acc    (conj acc diff)]
+      (if (or (seq more-expected)
+              (seq more-actual))
+        (recur acc more-expected more-actual)
+        (when (some some? acc)
+          acc)))))
```

Comment:
I'm just saying out loud that this doesn't necessarily enforce that `expected-seq` and `actual-seq` are the same length, but that e.g.

```
(=? [int? string? int?] [1 "two"])
```

will fail since `(=?-diff int? nil)` fails

Reply:
Also, `(=? [int? string? nil?] [1 "two"]))` succeeds, which might be somewhat counterintuitive. I can imagine whoever writes such an assertion wants to check if the actual value has at least three elements. 🤷